test_rclcpp/test/test_services_server.cpp

```diff
@@ -34,6 +34,27 @@ void handle_add_two_ints_reqid(
   response->sum = request->a + request->b;
 }
 
+class DeferedCbServiceWrapper
+{
+public:
+  explicit DeferedCbServiceWrapper(rclcpp::Node & node)
+  {
+    impl_ = node.create_service<test_rclcpp::srv::AddTwoInts>(
+      "add_two_ints_defered_cb",
+      [this](
+        const std::shared_ptr<rmw_request_id_t> request_header,
+        const std::shared_ptr<test_rclcpp::srv::AddTwoInts::Request> request)
+      {
+        test_rclcpp::srv::AddTwoInts::Response response;
+        response.sum = request->a + request->b;
+        this->impl_->send_response(*request_header, response);
```

Comment:
Should we do this from a thread or something? I'm just wondering if sending the response before the callback is done is some how hiding potential bugs. Also, it seems like users would want to send the response asynchronously after the callback, otherwise they'd just return the response (or set it from the arguments).

Reply:
> Should we do this from a thread or something? I'm just wondering if sending the response before the callback is done is some how hiding potential bugs. Also, it seems like users would want to send the response asynchronously after the callback, otherwise they'd just return the response (or set it from the arguments).

I agree that users are probably not going to use this this way, but I didn't want to make the test to complex.
Adding a thread sounds reasonable though

Reply:
See https://github.com/ros2/system_tests/pull/478/commits/81703c250838f76bb52c8aa463cb8eaa1297b669.